File: benchmark/src/nodejs/kmp_search.js
let seed = 42n;
function rng() {
    seed = (seed * 1103515245n + 12345n) % 2147483648n;
    return Number((seed - (seed % 65536n)) / 65536n) % 4;
}
const chars = "ACGT";
let textArr = [];
for (let i = 0; i < 500000; i++) textArr.push(chars[rng()]);
const text = textArr.join("");

const patterns = [];
for (let i = 0; i < 15; i++) {
    const startPos = Number((seed * 16807n) % 499000n);
    seed = (seed * 16807n) % 2147483647n;
    const plen = 5 + Number(seed % 11n);
    patterns.push(text.substring(startPos, startPos + plen));
}
for (let i = 0; i < 5; i++) {
    const pat = "QQQQQ" + String(Number((seed * 16807n) % 10n));
    seed = (seed * 16807n) % 2147483647n;
    patterns.push(pat);
}

const start = Date.now();
let totalMatches = 0;
for (const pat of patterns) {
    const m = pat.length;
    const fail = new Array(m).fill(0);
    let j = 0;
    for (let i = 1; i < m; i++) {
        while (j > 0 && pat[i] !== pat[j]) j = fail[j - 1];
        if (pat[i] === pat[j]) j++;
        fail[i] = j;
    }
    j = 0;
    for (let i = 0; i < text.length; i++) {
        while (j > 0 && text[i] !== pat[j]) j = fail[j - 1];
        if (text[i] === pat[j]) j++;
        if (j === m) { totalMatches++; j = fail[j - 1]; }
    }
}
const end = Date.now();
console.log(`Result: ${totalMatches}`);
console.log(`Time: ${end - start}ms`);

File: benchmark/src/python/kmp_search.py
import time

seed = 42
def rng():
    global seed
    seed = (seed * 1103515245 + 12345) % 2147483648
    return ((seed - (seed % 65536)) // 65536) % 4

chars = "ACGT"
text_chars = []
for _ in range(500000):
    text_chars.append(chars[rng()])
text = "".join(text_chars)

def get_seed():
    global seed
    return seed

patterns = []
for i in range(15):
    start_pos = (get_seed() * 16807) % 499000
    seed = (seed * 16807) % 2147483647
    plen = 5 + (seed % 11)
    pat = text[start_pos:start_pos + plen]
    patterns.append(pat)

for i in range(5):
    pat = "QQQQQ" + str((seed * 16807) % 10)
    seed = (seed * 16807) % 2147483647
    patterns.append(pat)

start = time.time() * 1000

total_matches = 0
for pat in patterns:
    m = len(pat)
    fail = [0] * m
    j = 0
    for i in range(1, m):
        while j > 0 and pat[i] != pat[j]:
            j = fail[j - 1]
        if pat[i] == pat[j]:
            j += 1
        fail[i] = j
    
    j = 0
    for i in range(len(text)):
        while j > 0 and text[i] != pat[j]:
            j = fail[j - 1]
        if text[i] == pat[j]:
            j += 1
        if j == m:
            total_matches += 1
            j = fail[j - 1]

end = time.time() * 1000
print(f"Result: {total_matches}")
print(f"Time: {int(end - start)}ms")
